gqcp/include/Processing/Properties/expectation_values.hpp

```diff
@@ -89,6 +88,46 @@ std::array<double, Components> calculateExpectationValue(const SQTwoElectronOper
 }
 
 
+/**
+ *  Calculate the expectation value for spin squared
+ * 
+ *  @tparam Scalar      the scalar type
+ * 
+ *  @param one_rdms        all the one electron density matrices
+ *  @param two_rdms        all the two electron density matrices
+ *
+ *  @return expectation value of spin squared
+ */
+template <typename Scalar>
+double calculateSpinSquared(const OneRDMs<Scalar>& one_rdms, const TwoRDMs<Scalar>& two_rdms) {
+    double sz = calculateSpinZ(one_rdms);
+    double s_squared = -sz;
+    size_t K = one_rdms.one_rdm.dimension();
+    for (size_t p = 0; p < K; p++) {
+        s_squared += one_rdms.one_rdm_aa(p, p) + (one_rdms.one_rdm_aa(p, p) +  one_rdms.one_rdm_bb(p, p))/4;
+        for (size_t q = 0; q < K; q++) {
+            s_squared += -two_rdms.two_rdm_aabb(p,q,q,p);
+            s_squared += (two_rdms.two_rdm_aaaa(p,p,q,q) + two_rdms.two_rdm_bbbb(p,p,q,q) - two_rdms.two_rdm_aabb(p,p,q,q) - two_rdms.two_rdm_bbaa(p,p,q,q))/4;
+        }
+    }
+    return s_squared;
```

Comment:
Are you calculating S^2 using S+S-, S_z^2 and S_z? If you are, could you indicate which portions of the code correspond to it?